R/step_climate.R

```diff
@@ -341,12 +350,63 @@ roll_modular_multivec <- function(col, .idx, weights, aggr, window_size, modulus
     tidyr::nest(data = c(col, weights), .by = .idx)
   out <- double(nrow(tib))
   for (iter in seq_along(out)) {
-    entries <- (iter - window_size):(iter + window_size) %% modulus
+    # +1 from 1-indexing
+    entries <- ((iter - window_size):(iter + window_size) %% modulus)
     entries[entries == 0] <- modulus
+    # note that because we are 1-indexing, we're looking for indices that are 1
+    # larger than the actual day/week in the year
+    if (modulus == 365) {
+      # we need to grab just the window around the leap day on the leap day
+      if (iter == 366) {
+        entries <- ((59 - window_size):(59 + window_size - 1) %% modulus)
+        entries <- c(entries, 366)
+      } else if ((59 %in% entries) || (60 %in% entries)) {
+        # if we're on the Feb/March boundary for daily data, we need to add in the
+        # leap day data
+        entries <- c(entries, 366)
+      }
+    } else if (modulus == 52) {
+      # we need to grab just the window around the leap week on the leap week
+      if (iter == 53) {
+        entries <- ((53 - window_size):(53 + window_size - 1) %% 52)
+        entries[entries == 0] <- 52
+        entries <- c(entries, 53)
+      } else if ((52 %in% entries) || (1 %in% entries)) {
+        # if we're on the year boundary for weekly data, we need to add in the
+        # leap week data (which is the extra week at the end)
+        entries <- c(entries, 53)
+      }
+    }
     out[iter] <- with(
       purrr::list_rbind(tib$data[entries]),
       aggr(col, weights)
     )
   }
   tibble(.idx = unique(tib$.idx), climate_pred = out)
 }
+
+
+#' a function that assigns Feb 29th to -1, and aligns all other dates the same
+#' number in the year, regardless of whether it's a leap year
+#' @keywords internal
+yday_leap <- function(time_value) {
+  ifelse(
+    !lubridate::leap_year(time_value),
+    lubridate::yday(time_value),
+    ifelse(
+      (lubridate::month(time_value) == 2) & lubridate::day(time_value) == 29,
+      999,
+      lubridate::yday(time_value) - (lubridate::month(time_value) > 2)
+    )
+  )
+}
+#' epiweek, but it assigns week 53 the value of -1 instead so it mirrors the assignments in yday_leap
```

Comment:
Do you mean this to say `-1` rather than `999`?

Reply:
yeah I initially wrote this using `-1` as the flag, but having the value at the end rather than the beginning proved somewhat useful.